src/ray/raylet/pubsub/pubsub_client.h

```diff
@@ -0,0 +1,138 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "ray/common/id.h"
+#include "ray/rpc/worker/core_worker_client_pool.h"
+#include "src/ray/protobuf/common.pb.h"
+
+namespace ray {
+
+using SubscriptionCallback = std::function<void(const ObjectID &)>;
+using SubscriptionFailureCallback = std::function<void(const ObjectID &)>;
+
+/// Interface for the pubsub client.
+class PubsubClientInterface {
```

Comment:
```suggestion
class SubscriberInterface {
```

Reply:
Technically this interface will support Publish methods as well in the future when it becomes more generic (like Redis client can publish stuff to the channel), so it is likely that we should change the name of the class to pubsub client again if we use Subscriber right now.

Simlarly, the PubsubCoordinator will be fully working a pubsub server (Since this will receives publish requests from pubsub clients and publish to other clients that subscribe the objects). 

I am not against the naming suggestion because the class is actually the subscriber now, and the pubsub coordinator is just a publisher. But I'd like you to know the reasoning of class names now. Lmk if you think we should not change the name to a Subscriber. (I will address your comments tonight).

Reply:
I think we would use a separate class for that right? I don't think it makes sense to be mixing both responsibilities into this.

Reply:
What is the use case for that? I don't see core worker as ever acting as a general pubsub server. It seems to me we only want:
- the owner publishing events about object status
- the GCS publishing events about node and actor status

Reply:
If we remove the Redis pubsub, we should implement our own pubsub (like for example, we need to publish error messages), and my expectation here is to use the same generalized pubsub classes to implement all that. I agree it is better not to over-generalizing this for now. I will just change the name then.

Reply:
I don't think we need generalized pub-sub... what would we use it for?

Reply:
I see. I thought we would replace Redis pubsub to a generalized pubsub impl on the GCS server (We use the pubsub for error messages, log messages, node information, actor information, job information, and potentially more parts that I am not aware of).

Reply:
Hmm that kind of makes sense, but really the GCS can internally generate all the messages it's publishing. There's no need to allow arbitrary publishing via GCS; only subscribing to a limited number of channels (errors, logs, nodes, actors, jobs). The "publishing" can be via specific proto APIs to GCS.

Reply:
Got it. That makes sense. I will rename the class as your suggestion.